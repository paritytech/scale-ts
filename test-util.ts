/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
