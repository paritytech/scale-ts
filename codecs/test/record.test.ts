import * as $ from "../../mod.ts"
import { testCodec, testInvalid } from "../../test-util.ts"

testCodec($.record($.str), [
  { a: "hi", b: "sup", c: "yo" },
  { the: "quick", brown: "fox", jumped: "over", theLazy: "hen" },
])

testCodec($.record($.record($.u8)), [{
  one: { two: 3 },
  four: { five: 6 },
  seven: { eight: 9 },
}])

testInvalid($.record($.str), [[true, false], [1, 2, 3, -1, 4]])
testInvalid($.record($.u8), [{ this: "should" }, { be: "invalid" }, { and: "this" }])
