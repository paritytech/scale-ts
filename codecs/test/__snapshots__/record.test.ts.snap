export const snapshot = {};

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.str)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) { a: "hi", b: "sup", c: "yo" } 1`] = `
0c
04
61
08
68
69
04
62
0c
73
75
70
04
63
08
79
6f
`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.str)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) { the: "quick", brown: "fox", jumped: "over", theLazy: "hen" } 1`] = `
10
0c
74
68
65
14
71
75
69
63
6b
14
62
72
6f
77
6e
0c
66
6f
78
18
6a
75
6d
70
65
64
10
6f
76
65
72
1c
74
68
65
4c
61
7a
79
0c
68
65
6e
`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.transform(Array))),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) { one: { two: 3 }, four: { five: 6 }, seven: { eight: 9 } } 1`] = `
0c
0c
6f
6e
65
04
0c
74
77
6f
03
10
66
6f
75
72
04
10
66
69
76
65
06
14
73
65
76
65
6e
04
14
65
69
67
68
74
09
`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.str)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) invalid [ true, false ] 1`] = `ScaleAssertError: typeof value#encode[0][1] !== "string"`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.str)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) invalid [ 1, 2, 3, -1, 4 ] 1`] = `ScaleAssertError: typeof value#encode[0][1] !== "string"`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.u8)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) invalid { this: "should" } 1`] = `ScaleAssertError: typeof value#encode[0][1] !== "number"`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.u8)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) invalid { be: "invalid" } 1`] = `ScaleAssertError: typeof value#encode[0][1] !== "number"`;

snapshot[`\$.transform(
  {
    "\$base": \$.array(\$.tuple(\$.str, \$.u8)),
    encode: [Function: entries],
    decode: [Function: fromEntries]
  }
) invalid { and: "this" } 1`] = `ScaleAssertError: typeof value#encode[0][1] !== "number"`;
